### PR TITLE
CFE-965: [V2] Enable signature verification

### PR DESCRIPTION
**Pull request by lmzuccarelli**
# Description

This change enables secure policy for signature verification on release and additionalImages (in V2) for all RedHat based images

Refer to v2/README.md for more details

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Testing 
-------

Checking additional images first - imagesetconfig

```yaml
kind: ImageSetConfiguration
apiVersion: mirror.openshift.io/v1alpha2
storageConfig:
  local:
    path: /tmp/storageBackend

mirror:
  additionalImages:
  - name: registry.access.redhat.com/ubi8/ubi:latest

```

Ensure the additionalImages signature verification fails

Set a bad entry in /etc/container/policy.json (for registry.access.redhat.com)

```json
{
  "default": [
      {
          "type": "insecureAcceptAnything"
      }
  ],
  "transports":
    {
      "docker-daemon":
          {
              "": [{"type":"insecureAcceptAnything"}]
          },
      "docker":
        {
          "registry.redhat.io/redhat/certified-operator-index": [
            {
              "type": "signedBy",
              "keyType": "GPGKeys",
              "keyPath": "/etc/pki/rpm-gpg/RPM-GPG-KEY-redhat-isv"
            }
          ],
          "registry.redhat.io/redhat/community-operator-index": [
            {
              "type": "signedBy",
              "keyType": "GPGKeys",
              "keyPath": "/etc/pki/rpm-gpg/RPM-GPG-KEY-redhat-isv"
            }
          ],
          "registry.redhat.io/redhat/redhat-marketplace-index": [
            {
              "type": "signedBy",
              "keyType": "GPGKeys",
              "keyPath": "/etc/pki/rpm-gpg/RPM-GPG-KEY-redhat-isv"
            }
          ],
          "registry.redhat.io": [
            {
              "type": "signedBy",
              "keyType": "GPGKeys",
              "keyPath": "/etc/pki/rpm-gpg/RPM-GPG-KEY-fedora-9-primary"
            }
          ],
	        "registry.access.redhat.com": [
            {
              "type": "signedBy",
              "keyType": "GPGKeys",
              "keyPath": "/etc/pki/rpm-gpg/RPM-GPG-KEY-fedora-9-primary"
            }
          ],
	        "quay.io/openshift-release-dev/openshift-release-dev" :[
 	          {
              "type": "signedBy",
              "keyType": "GPGKeys",
              "keyPath": "/etc/pki/rpm-gpg/RPM-GPG-KEY-redhat-release"
            }
	        ]
        }
    }
}

```

Execute oc-binary with --secure-policy flag

```bash
oc-mirror --v2 --config imstcfg.yaml file://cfe-965 --secure-policy

```

Console output

```bash
2023/10/12 13:53:38  [INFO]   : mode mirrorToDisk 
2023/10/12 13:53:38  [INFO]   : local storage registry will log to /home/lzuccarelli/go/src/github.com/openshift/oc-mirror/logs/registry.log
2023/10/12 13:53:38  [INFO]   : starting local storage on :5000
2023/10/12 13:53:38  [INFO]   : total release images to copy 0 
2023/10/12 13:53:38  [INFO]   : related images length 0 
2023/10/12 13:53:38  [INFO]   : images to copy (before duplicates) 0 
2023/10/12 13:53:38  [INFO]   : total operator images to copy 0 
2023/10/12 13:53:38  [INFO]   : total additional images to copy 1 
2023/10/12 13:53:38  [INFO]   : images to mirror 1 
2023/10/12 13:53:38  [INFO]   : batch count 1 
2023/10/12 13:53:38  [INFO]   : batch index 0 
2023/10/12 13:53:38  [INFO]   : batch size 1 
2023/10/12 13:53:38  [INFO]   : remainder size 0 
2023/10/12 13:53:38  [INFO]   : starting batch 0 
2023/10/12 13:53:41  [INFO]   : completed batch 0
2023/10/12 13:53:41  [ERROR]  : [Worker] errArray copying image 1/4 from manifest list: Source image rejected: None of the signatures were accepted, reasons: Invalid GPG signature: gpgme.Signature{Summary:128, Fingerprint:"199E2F91FD431D51", Status:gpgme.Error{err:0x9}, Timestamp:time.Date(2023, time.October, 9, 11, 31, 34, 0, time.Local), ExpTimestamp:time.Date(1970, time.January, 1, 1, 0, 0, 0, time.Local), WrongKeyUsage:false, PKATrust:0x0, ChainModel:false, Validity:0, ValidityReason:error(nil), PubkeyAlgo:1, HashAlgo:8}; Invalid GPG signature: gpgme.Signature{Summary:128, Fingerprint:"199E2F91FD431D51", Status:gpgme.Error{err:0x9}, Timestamp:time.Date(2023, time.October, 9, 11, 31, 34, 0, time.Local), ExpTimestamp:time.Date(1970, time.January, 1, 1, 0, 0, 0, time.Local), WrongKeyUsage:false, PKATrust:0x0, ChainModel:false, Validity:0, ValidityReason:error(nil), PubkeyAlgo:1, HashAlgo:8}; Invalid GPG signature: gpgme.Signature{Summary:128, Fingerprint:"199E2F91FD431D51", Status:gpgme.Error{err:0x9}, Timestamp:time.Date(2023, time.October, 9, 11, 31, 34, 0, time.Local), ExpTimestamp:time.Date(1970, time.January, 1, 1, 0, 0, 0, time.Local), WrongKeyUsage:false, PKATrust:0x0, ChainModel:false, Validity:0, ValidityReason:error(nil), PubkeyAlgo:1, HashAlgo:8}; Invalid GPG signature: gpgme.Signature{Summary:128, Fingerprint:"199E2F91FD431D51", Status:gpgme.Error{err:0x9}, Timestamp:time.Date(2023, time.October, 9, 11, 31, 34, 0, time.Local), ExpTimestamp:time.Date(1970, time.January, 1, 1, 0, 0, 0, time.Local), WrongKeyUsage:false, PKATrust:0x0, ChainModel:false, Validity:0, ValidityReason:error(nil), PubkeyAlgo:1, HashAlgo:8}; Invalid GPG signature: gpgme.Signature{Summary:128, Fingerprint:"199E2F91FD431D51", Status:gpgme.Error{err:0x9}, Timestamp:time.Date(2023, time.October, 9, 11, 31, 34, 0, time.Local), ExpTimestamp:time.Date(1970, time.January, 1, 1, 0, 0, 0, time.Local), WrongKeyUsage:false, PKATrust:0x0, ChainModel:false, Validity:0, ValidityReason:error(nil), PubkeyAlgo:1, HashAlgo:8}; Invalid GPG signature: gpgme.Signature{Summary:128, Fingerprint:"199E2F91FD431D51", Status:gpgme.Error{err:0x9}, Timestamp:time.Date(2023, time.October, 9, 11, 31, 34, 0, time.Local), ExpTimestamp:time.Date(1970, time.January, 1, 1, 0, 0, 0, time.Local), WrongKeyUsage:false, PKATrust:0x0, ChainModel:false, Validity:0, ValidityReason:error(nil), PubkeyAlgo:1, HashAlgo:8}
2023/10/12 13:53:41  [INFO]   : start time: 2023-10-12 13:53:38.936597935 +0200 CEST m=+0.068315660
collection time: 2023-10-12 13:53:38.936887769 +0200 CEST m=+0.068605489
mirror time: 2023-10-12 13:53:41.276122489 +0200 CEST m=+2.407840215
2023/10/12 13:53:41  [ERROR]  : [Worker] error in batch - refer to console logs

```

Update /etc/containers/policy.json to correct value

```json
{
  "default": [
      {
          "type": "insecureAcceptAnything"
      }
  ],
  "transports":
    {
      "docker-daemon":
          {
              "": [{"type":"insecureAcceptAnything"}]
          },
      "docker":
        {
          "registry.redhat.io/redhat/certified-operator-index": [
            {
              "type": "signedBy",
              "keyType": "GPGKeys",
              "keyPath": "/etc/pki/rpm-gpg/RPM-GPG-KEY-redhat-isv"
            }
          ],
          "registry.redhat.io/redhat/community-operator-index": [
            {
              "type": "signedBy",
              "keyType": "GPGKeys",
              "keyPath": "/etc/pki/rpm-gpg/RPM-GPG-KEY-redhat-isv"
            }
          ],
          "registry.redhat.io/redhat/redhat-marketplace-index": [
            {
              "type": "signedBy",
              "keyType": "GPGKeys",
              "keyPath": "/etc/pki/rpm-gpg/RPM-GPG-KEY-redhat-isv"
            }
          ],
          "registry.redhat.io": [
            {
              "type": "signedBy",
              "keyType": "GPGKeys",
              "keyPath": "/etc/pki/rpm-gpg/RPM-GPG-KEY-redhat-release"
            }
          ],
	        "registry.access.redhat.com": [
            {
              "type": "signedBy",
              "keyType": "GPGKeys",
              "keyPath": "/etc/pki/rpm-gpg/RPM-GPG-KEY-redhat-release"
            }
          ],
	        "quay.io/openshift-release-dev/openshift-release-dev" :[
 	          {
              "type": "signedBy",
              "keyType": "GPGKeys",
              "keyPath": "/etc/pki/rpm-gpg/RPM-GPG-KEY-redhat-release"
            }
	        ]
        }
    }
}

```

Re-run oc-mirror

```bash

oc-mirror --v2 --config imstcfg.yaml file://cfe-965 --secure-policy

```

Console output

```bash

2023/10/12 14:04:51  [INFO]   : mode mirrorToDisk
2023/10/12 14:04:51  [INFO]   : local storage registry will log to /home/lzuccarelli/go/src/github.com/openshift/oc-mirror/logs/registry.log
2023/10/12 14:04:51  [INFO]   : starting local storage on :5000
2023/10/12 14:04:51  [INFO]   : total release images to copy 0
2023/10/12 14:04:51  [INFO]   : related images length 0
2023/10/12 14:04:51  [INFO]   : images to copy (before duplicates) 0
2023/10/12 14:04:51  [INFO]   : total operator images to copy 0
2023/10/12 14:04:51  [INFO]   : total additional images to copy 1
2023/10/12 14:04:51  [INFO]   : images to mirror 1
2023/10/12 14:04:51  [INFO]   : batch count 1
2023/10/12 14:04:51  [INFO]   : batch index 0
2023/10/12 14:04:51  [INFO]   : batch size 1
2023/10/12 14:04:51  [INFO]   : remainder size 0
2023/10/12 14:04:51  [INFO]   : starting batch 0
2023/10/12 14:07:44  [INFO]   : completed batch 0
2023/10/12 14:07:44  [INFO]   : start time: 2023-10-12 14:04:51.824126921 +0200 CEST m=+0.056019003
collection time: 2023-10-12 14:04:51.824638937 +0200 CEST m=+0.056531021
mirror time: 2023-10-12 14:07:44.201467392 +0200 CEST m=+172.433359478

```

Update imagesetconfig to include a release

```yaml

kind: ImageSetConfiguration
apiVersion: mirror.openshift.io/v1alpha2
storageConfig:
  local:
    path: /tmp/storageBackend

mirror:
  platform:
    channels:
    - name: stable-4.13
      type: ocp
      minVersion: 4.13.1
      maxVersion: 4.1i3.1

```

Ensure the release signature fails 

Update /etc/containers/policy.json with a bad entry

```bash
{
  "default": [
      {
          "type": "insecureAcceptAnything"
      }
  ],
  "transports":
    {
      "docker-daemon":
          {
              "": [{"type":"insecureAcceptAnything"}]
          },
      "docker":
        {
          "registry.redhat.io/redhat/certified-operator-index": [
            {
              "type": "signedBy",
              "keyType": "GPGKeys",
              "keyPath": "/etc/pki/rpm-gpg/RPM-GPG-KEY-redhat-isv"
            }
          ],
          "registry.redhat.io/redhat/community-operator-index": [
            {
              "type": "signedBy",
              "keyType": "GPGKeys",
              "keyPath": "/etc/pki/rpm-gpg/RPM-GPG-KEY-redhat-isv"
            }
          ],
          "registry.redhat.io/redhat/redhat-marketplace-index": [
            {
              "type": "signedBy",
              "keyType": "GPGKeys",
              "keyPath": "/etc/pki/rpm-gpg/RPM-GPG-KEY-redhat-isv"
            }
          ],
          "registry.redhat.io": [
            {
              "type": "signedBy",
              "keyType": "GPGKeys",
              "keyPath": "/etc/pki/rpm-gpg/RPM-GPG-KEY-redhat-release"
            }
          ],
	  "registry.access.redhat.com": [
            {
              "type": "signedBy",
              "keyType": "GPGKeys",
              "keyPath": "/etc/pki/rpm-gpg/RPM-GPG-KEY-redhat-release"
            }
          ],
	  "quay.io/openshift-release-dev/openshift-release-dev" :[
 	    {
              "type": "signedBy",
              "keyType": "GPGKeys",
              "keyPath": "/etc/pki/rpm-gpg/RPM-GPG-KEY-fedora-39-primary"
            }
	  ]
        }
    }
}

```

Execute oc-mirror

```bash

oc-mirror --v2 --config imstcfg.yaml file://cfe-965 --secure-policy

```

Console output

```bash
2023/10/12 14:18:44  [INFO]   : mode mirrorToDisk 
2023/10/12 14:18:44  [INFO]   : local storage registry will log to /home/lzuccarelli/go/src/github.com/openshift/oc-mirror/logs/registry.log
2023/10/12 14:18:44  [INFO]   : starting local storage on :5000
2023/10/12 14:18:44  [INFO]   : processing minimum version 4.13.1 and maximum version 4.13.1
2023/10/12 14:18:45  [INFO]   : Found update 4.13.1
2023/10/12 14:18:45  [WARN]   : signature for 9c92b5ec203ee7f81626cc4e9f02086484056a76548961e5895916f136302b1f not in cache
2023/10/12 14:18:45  [INFO]   : signature 9c92b5ec203ee7f81626cc4e9f02086484056a76548961e5895916f136302b1f
2023/10/12 14:18:46  [INFO]   : content {"critical": {"image": {"docker-manifest-digest": "sha256:9c92b5ec203ee7f81626cc4e9f02086484056a76548961e5895916f136302b1f"}, "type": "atomic container signature", "identity": {"docker-reference": "quay.io/openshift-release-dev/ocp-release:4.13.1-x86_64"}}, "optional": {"creator": "Red Hat OpenShift Signing Authority 0.0.1"}}
2023/10/12 14:18:46  [INFO]   : image found : quay.io/openshift-release-dev/ocp-release:4.13.1-x86_64
2023/10/12 14:18:46  [INFO]   : public Key : 567E347AD0044ADE55BA8A5F199E2F91FD431D51
2023/10/12 14:18:46  [INFO]   : copying  quay.io/openshift-release-dev/ocp-release:4.13.1-x86_64 
2023/10/12 14:19:45  [INFO]   : total release images to copy 184 
2023/10/12 14:19:45  [INFO]   : related images length 0 
2023/10/12 14:19:45  [INFO]   : images to copy (before duplicates) 0 
2023/10/12 14:19:45  [INFO]   : total operator images to copy 0 
2023/10/12 14:19:45  [INFO]   : total additional images to copy 0 
2023/10/12 14:19:45  [INFO]   : images to mirror 184 
2023/10/12 14:19:45  [INFO]   : batch count 23 
2023/10/12 14:19:45  [INFO]   : batch index 0 
2023/10/12 14:19:45  [INFO]   : batch size 8 
2023/10/12 14:19:45  [INFO]   : remainder size 0 
2023/10/12 14:19:45  [INFO]   : starting batch 0 
2023/10/12 14:19:47  [INFO]   : completed batch 0
2023/10/12 14:19:47  [ERROR]  : [Worker] errArray initializing source docker://quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:1dac5e9e31c011322d414a74f1aeb1ab2b4d766679b2ea7df359fa3456a7b83d: reading manifest sha256:1dac5e9e31c011322d414a74f1aeb1ab2b4d766679b2ea7df359fa3456a7b83d in quay.io/openshift-release-dev/ocp-v4.0-art-dev: unauthorized: access to the requested resource is not authorized
2023/10/12 14:19:47  [ERROR]  : [Worker] errArray initializing source docker://quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:3a8b33263729ab42c0ff29b9d5e8b767b7b1a9b31240c592fa8d173463fb04d1: reading manifest sha256:3a8b33263729ab42c0ff29b9d5e8b767b7b1a9b31240c592fa8d173463fb04d1 in quay.io/openshift-release-dev/ocp-v4.0-art-dev: unauthorized: access to the requested resource is not authorized
2023/10/12 14:19:47  [ERROR]  : [Worker] errArray initializing source docker://quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:2e1b327b96ce3b897f25133b0d119fbe671e499dee2483e72f5953cc85947a06: reading manifest sha256:2e1b327b96ce3b897f25133b0d119fbe671e499dee2483e72f5953cc85947a06 in quay.io/openshift-release-dev/ocp-v4.0-art-dev: unauthorized: access to the requested resource is not authorized
2023/10/12 14:19:47  [ERROR]  : [Worker] errArray initializing source docker://quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:ce3e2e4aac617077ac98b82d9849659595d85cd31f17b3213da37bc5802b78e1: reading manifest sha256:ce3e2e4aac617077ac98b82d9849659595d85cd31f17b3213da37bc5802b78e1 in quay.io/openshift-release-dev/ocp-v4.0-art-dev: unauthorized: access to the requested resource is not authorized
2023/10/12 14:19:47  [ERROR]  : [Worker] errArray initializing source docker://quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:c9836bc7472215104f20ba7f0656ad69061681cc626165cee18470645c880d33: reading manifest sha256:c9836bc7472215104f20ba7f0656ad69061681cc626165cee18470645c880d33 in quay.io/openshift-release-dev/ocp-v4.0-art-dev: unauthorized: access to the requested resource is not authorized
2023/10/12 14:19:47  [ERROR]  : [Worker] errArray initializing source docker://quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:ee080b37108cc5bee6b0e0d1e55148f4aa67e3d236d53fb7d9aee4a83d187603: reading manifest sha256:ee080b37108cc5bee6b0e0d1e55148f4aa67e3d236d53fb7d9aee4a83d187603 in quay.io/openshift-release-dev/ocp-v4.0-art-dev: unauthorized: access to the requested resource is not authorized
2023/10/12 14:19:47  [ERROR]  : [Worker] errArray initializing source docker://quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:29ef5c21337a918bb51939045206386790d22dec33cc705f668bdd0c7c9d5b51: reading manifest sha256:29ef5c21337a918bb51939045206386790d22dec33cc705f668bdd0c7c9d5b51 in quay.io/openshift-release-dev/ocp-v4.0-art-dev: unauthorized: access to the requested resource is not authorized
2023/10/12 14:19:47  [ERROR]  : [Worker] errArray initializing source docker://quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:6a6bb6f9a57bc5484a3c3e3223606719c963b9939b07d40c2c7eaccf97ce6a20: reading manifest sha256:6a6bb6f9a57bc5484a3c3e3223606719c963b9939b07d40c2c7eaccf97ce6a20 in quay.io/openshift-release-dev/ocp-v4.0-art-dev: unauthorized: access to the requested resource is not authorized
2023/10/12 14:19:47  [INFO]   : start time: 2023-10-12 14:18:44.486017397 +0200 CEST m=+0.063208194
collection time: 2023-10-12 14:19:45.331170391 +0200 CEST m=+60.908361187
mirror time: 2023-10-12 14:19:47.390820731 +0200 CEST m=+62.968011585
2023/10/12 14:19:47  [ERROR]  : [Worker] error in batch - refer to console logs

```

Update the /etc/containers/policy.json to the correct value for quay.io/openshift-release-dev/openshift-release-dev

```bash
{
  "default": [
      {
          "type": "insecureAcceptAnything"
      }
  ],
  "transports":
    {
      "docker-daemon":
          {
              "": [{"type":"insecureAcceptAnything"}]
          },
      "docker":
        {
          "registry.redhat.io/redhat/certified-operator-index": [
            {
              "type": "signedBy",
              "keyType": "GPGKeys",
              "keyPath": "/etc/pki/rpm-gpg/RPM-GPG-KEY-redhat-isv"
            }
          ],
          "registry.redhat.io/redhat/community-operator-index": [
            {
              "type": "signedBy",
              "keyType": "GPGKeys",
              "keyPath": "/etc/pki/rpm-gpg/RPM-GPG-KEY-redhat-isv"
            }
          ],
          "registry.redhat.io/redhat/redhat-marketplace-index": [
            {
              "type": "signedBy",
              "keyType": "GPGKeys",
              "keyPath": "/etc/pki/rpm-gpg/RPM-GPG-KEY-redhat-isv"
            }
          ],
          "registry.redhat.io": [
            {
              "type": "signedBy",
              "keyType": "GPGKeys",
              "keyPath": "/etc/pki/rpm-gpg/RPM-GPG-KEY-redhat-release"
            }
          ],
	        "registry.access.redhat.com": [
            {
              "type": "signedBy",
              "keyType": "GPGKeys",
              "keyPath": "/etc/pki/rpm-gpg/RPM-GPG-KEY-redhat-release"
            }
          ],
	        "quay.io/openshift-release-dev/openshift-release-dev" :[
 	          {
              "type": "signedBy",
              "keyType": "GPGKeys",
              "keyPath": "/etc/pki/rpm-gpg/RPM-GPG-KEY-redhat-release"
            }
	        ]
        }
    }
}

```

This should now execute without any problems

# Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: v2/pkg/cli/executor.go [hunk 1]
```diff
@@ -331,6 +330,7 @@ func (o *ExecutorSchema) Complete(args []string) {
 	} else {
 		o.Log.Error("unable to determine the mode (the destination must be either file:// or docker://)")
 	}
+	//o.Opts.Global.SecurePolicy = false
```
Comment (sherine-k) [inline v2/pkg/cli/executor.go:333]: Nit: can be removed

File: v2/pkg/cli/executor.go [hunk 2]
```diff
@@ -361,12 +361,6 @@ func (o *ExecutorSchema) Run(cmd *cobra.Command, args []string) error {
 	}
 
 	if o.Opts.Mode == mirrorToDisk {
-		// ensure working dir exists
-		err := os.MkdirAll(workingDir, 0755)
-		if err != nil {
-			o.Log.Error(" %v ", err)
-			return err
-		}
```
Comment (sherine-k) [inline v2/pkg/cli/executor.go:369]: I thought it was a good idea to ensure that workingDir existed before starting. Is this being initialized elsewhere?

**Review by sherine-k — COMMENTED** [review]
Nice work @lmzuccarelli ! 
Thank you
I have just a small doubt about removing the creation of `workingDir` in executor.go code. 
LGTM otherwise ;)

File: v2/pkg/cli/executor.go (hunk 2 above)
Comment (lmzuccarelli) [reply v2/pkg/cli/executor.go]: I removed it as v2 does not use working dir - but just for sanity sake I added it back again :)

**Review by sherine-k — APPROVED** [review]
/lgtm

File: v2/pkg/diff/diff.go [hunk 3]
```diff
@@ -1,268 +0,0 @@
-package diff
```
Comment (aguidirh) [inline v2/pkg/diff/diff.go:1]: Good to see this diff mechanism going away

**Review by aguidirh — COMMENTED** [review]
/lgtm

**Pull request merged**